### ISSUE TITLE
[TEST] startTLS is broken if there's already data in the outgoing buffer.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10445]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#686|
|Type|defect|
|Created|2004-08-24 05:53:34Z|
|Branch|https://github.com/twisted/twisted/tree/starttls-with-buffered-686-3|

<details><summary>Searchable metadata</summary>

```
trac-id__686 686
type__defect defect
reporter__jknight jknight
priority__high high
milestone__ 
branch__branches_starttls_with_buffered_686_3 branches/starttls-with-buffered-686-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1093326814000000 1093326814000000
changetime__1218381583000000 1218381583000000
version__ 
owner__ 
cc__exarkun cc__itamarst cc__jknight cc__thijs cc__therve cc__pjd
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
If you do write("lotsofdata"); startTLS(), sometimes twisted will do the startTLS() before the 
write("lotsofdata"). The condition for "sometimes" being whether the lotsofdata was large enough to 
cause the data to be buffered in twisted, instead of sent immediately, or buffered by the kernel.

Apparently this doesn't show up normally, because people don't often send much data before doing 
STARTTLS, and thus the data normally *is* sent out immediately.

See also twisted.test.test_ssl.SpammyTLSTestCase test cases.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Propose this patch to make the problem more visible and less data-corruptful if you do happen to run 
into it:

Index: twisted/internet/tcp.py
===============================================================
====
--- twisted/internet/tcp.py     (revision 11439)
+++ twisted/internet/tcp.py     (working copy)
@@ -211,6 +211,17 @@

         def startTLS(self, ctx):
             assert not self.TLS
+            if self.dataBuffer:
+                written = self.writeSomeData(buffer(self.dataBuffer, self.offset))
+                offset = self.offset
+                dataLen = len(self.dataBuffer)
+                self.offset = 0
+                self.dataBuffer = ""
+                if isinstance(written, Exception) or (offset + written != len(dataBuffer)):
+                    warnings.warn("startTLS with unwritten buffered data currently doesn't work right. See 
issue #686. Closing connection.", category=RuntimeWarning, stacklevel=2)
+                    self.loseConnection()
+
+
             self.stopReading()
             self.stopWriting()
             self._startTLS()
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
This should be fixed correctly by 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Still urgent, but jknight explained that it's very unlikely to happen in
practice given his fixes, so it's not necessary for 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
urgent *means* "next release", so it's not urgent (not to mention no failing
unit test), sorry :-)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
twisted.test.test_ssl.SpammyTLSTestCase covers this bug, I think.  Note that it
is set todo.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_6):
> twisted.test.test_ssl.SpammyTLSTestCase covers this bug, I think.  Note that it
> is set todo.

I don't see any TODOs in the [twisted/test/test_ssl.py SpammyTLSTestCase](twisted/test/test_ssl.py SpammyTLSTestCase) anymore, guess this means it's fixed?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set owner to @jyknight|
|-|-|

Replying to [thijs](#note_7):
> I don't see any TODOs in the [twisted/test/test_ssl.py SpammyTLSTestCase](twisted/test/test_ssl.py SpammyTLSTestCase) anymore, guess this means it's fixed?

Oh damn, I now see the :( faces and todo properties.. sorry for the noise.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [24382]) Branching to 'starttls-with-buffered-686'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

I think this is basically resolved, but it's blocked on [#2874](https://github.com/twisted/twisted/issues/2874).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [24414]) Branching to 'starttls-with-buffered-686-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @exarkun|
|-|-|

Great, nice fix. Some small comments:

Flakes:
```
twisted/test/test_ssl.py:8: 'trial_util' imported but unused
twisted/internet/tcp.py:20: 'warnings' imported but unused
```

```
+                self._tlsBufferedData = []
+                self._tlsContext = ctx
+                self._tlsExtra = extra
```
I think it may be better if one object could hold all this state, eg `self._tlsDelayed = _TLSDelayed([], ctx, extra)`. What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [24509]) fix pyflakes warnings

refs [#10445](https://github.com/twisted/twisted/issues/10445)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [24510]) replace four _tls* attributes with _tlsWaiting

refs [#10445](https://github.com/twisted/twisted/issues/10445)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun removed owner|
|-|-|

I've made those two changes in the two revisions linked above (no other changes since the last review).  Here are the most current build results:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/starttls-with-buffered-686-2

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set owner to @exarkun|
|-|-|

I'm ok with the branch, but the build results show some problems:
 * gtk2/glib2 is not happy on windows it seems
 * I've noticed that the runReactor method doesn't cancel the timeout :/ Should be pretty easy to fix, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [24525]) Branching to 'starttls-with-buffered-686-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @therve|
|-|-|

I think those points have been taken care of in trunk already.  I merged forward, here are the results for the new branch:

  http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/starttls-with-buffered-686-3

I should have done that before, I knew there were still weird problems in the -2 branch.

`runReactor` still doesn't _directly_ cancel its timeout, but `unbuildReactor` does.  I think this is good enough until we have a case that demands something more.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [24527]) Merge starttls-with-buffered-686-3

Author: exarkun
Reviewer: jknight, therve
Fixes: [#10445](https://github.com/twisted/twisted/issues/10445)

Change the implementation of ITLSTransport.startTLS so that
if there is unsent data in the buffer, TLS negotiation is
delayed until all of that data is sent.  This changes the
existing behavior which would log an error if the data
could not be sent immediately and then drop the connection.

**Comment by twisted-trac** [comment]
|<img alt="pjd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|pjd commented|
|-|-|

Note that this has some fallout:  see Divmod ticket [http://divmod.org/trac/ticket/2673 [#2673](https://github.com/twisted/twisted/issues/2673)].

Callers can no longer safely assume that the TLS session is established once `startTLS` returns, and code that used to work due to `startTLS` opportunistically flushing the buffer can suddenly start failing.

I'm guessing that `startTLS` should be defined to return a `Deferred` (or provide some other mechanism) to allow callers to wait for the delayed TLS session.

(Should this be reopened?)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

One small correction: callers could never assume that the TLS session was established immediately after `startTLS` returned; they could rely on one of two things - that the connection was closed if there was data in the application buffer which couldn't be sent, or that the TLS handshake had probably started.  As it turns out, the handle returned by `getHandle` would also be a pyOpenSSL `Connection` object, but the `getHandle` documentation is pretty clear:

_Return a system- and reactor-specific handle.  This might be a socket.socket() object, or some other type of object, depending on which reactor is being used. **Use and manipulate at your own risk**.  This might be used in cases where you want to set specific options not exposed by the Twisted APIs._

I consider the change made here, from always returning a pyOpenSSL `Connection` object immediately after a `startTLS` call returning one at a somewhat later time in some cases to be reasonable.  I also don't expect that it will break anything except for `problemsFromTransport` (however, if it does break some other applications, then I'll happily consider making some further changes to avoid that).

I agree that there should be a callback to notify application code of successful (and likely failed) TLS initiation, but that's a task for a separate ticket.  Additionally, in all real-world protocols that I'm aware of, such callbacks aren't particularly useful, since the protocol is designed to allow that information to be discovered in other ways.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
